### PR TITLE
kafka(ticdc): add factory interface 

### LINKED ISSUE #8213: introduce factory interface to manage all kafka components

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve?

Issue Number: close #8213 

### What is changed and how it works?
* add a new interface, named `kafka.Factory` to manage all kafka components, it's an abstract factory, implemented by `sarama`, `kafka-go` and `mock`.
* move some sarama specific logic into sarama factory's implementation, especially the aysnc close
* move some shared methods from sink v1 into `pkg/sink/kafka`

#### Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

### Release note 

```release-note
`None`
```

### CODE REVIEW

File: cdc/sinkv2/eventsink/mq/dmlproducer/dml_producer.go [hunk 1]
```diff
@@ -37,5 +37,5 @@ type DMLProducer interface {
 // there is no way to safely close errCh by the sender.
 // So we let the GC close errCh.
 // It's usually a buffered channel.
-type Factory func(ctx context.Context, client kafka.Client,
+type Factory func(ctx context.Context, client kafka.Factory,
```
Comment (3AceShowHand) [inline cdc/sinkv2/eventsink/mq/dmlproducer/dml_producer.go:40]: rename to factory

File: pkg/sink/kafka/metrics_collector.go [hunk 2]
```diff
@@ -52,7 +51,8 @@ type saramaMetricsCollector struct {
 	// adminClient is used to get broker infos from broker.
 	adminClient ClusterAdminClient
 	brokers     map[int32]struct{}
-	registry    metrics.Registry
+
+	registry metrics.Registry
```
Comment (3AceShowHand) [inline pkg/sink/kafka/metrics_collector.go:55]: remove blank

**Review by sdojjy — APPROVED** [review]

File: cdc/sinkv2/ddlsink/mq/mq_ddl_sink.go [hunk 3]
```diff
@@ -54,10 +55,14 @@ type ddlSink struct {
 	producer ddlproducer.DDLProducer
 	// statistics is used to record DDL metrics.
 	statistics *metrics.Statistics
+
```
Comment (zhaoxinyu) [inline cdc/sinkv2/ddlsink/mq/mq_ddl_sink.go:58]:
```suggestion
```

File: cdc/sinkv2/ddlsink/mq/mq_ddl_sink.go [hunk 4]
```diff
@@ -175,4 +181,12 @@ func (k *ddlSink) Close() {
 	if k.producer != nil {
 		k.producer.Close()
 	}
+	if k.admin != nil {
+		if err := k.admin.Close(); err != nil {
```
Comment (zhaoxinyu) [inline cdc/sinkv2/ddlsink/mq/mq_ddl_sink.go:185]: What would happen if we don't throw this error to caller? Will the error neglect cause problems like goroutine leak or something else?

File: cdc/sinkv2/eventsink/mq/dmlproducer/kafka_dml_producer.go [hunk 5]
```diff
@@ -70,35 +66,16 @@ func NewKafkaDMLProducer(
 
 	closeCh := make(chan struct{})
 	failpointCh := make(chan error, 1)
-	asyncProducer, err := client.AsyncProducer(changefeedID, closeCh, failpointCh)
+	asyncProducer, err := factory.AsyncProducer(closeCh, failpointCh)
 	if err != nil {
-		// Close the client to prevent the goroutine leak.
```
Comment (zhaoxinyu) [inline cdc/sinkv2/eventsink/mq/dmlproducer/kafka_dml_producer.go:75]: Ditto. The removal of async close logic

File: pkg/sink/kafka/sarama_factory.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+		brokerEndpoints: o.BrokerEndpoints,
+		config:          saramaConfig,
+	}, nil
+}
+
+func (f *saramaFactory) AdminClient() (ClusterAdminClient, error) {
+	return newAdminClient(f.brokerEndpoints, f.config)
+}
+
+// SyncProducer return a Sync Producer,
```
Comment (zhaoxinyu) [inline pkg/sink/kafka/sarama_factory.go:51]:
```suggestion
// SyncProducer returns a Sync Producer,
```

File: cdc/sinkv2/ddlsink/mq/ddlproducer/kafka_ddl_producer.go [hunk 7]
```diff
@@ -153,47 +129,8 @@ func (k *kafkaDDLProducer) Close() {
 		return
 	}
 	k.closed = true
-	// We need to close it asynchronously. Otherwise, we might get stuck
```
[the remark this answers is not in the document]
Comment (3AceShowHand) [reply cdc/sinkv2/ddlsink/mq/ddlproducer/kafka_ddl_producer.go:156]: It's moved inside `sarama.syncProducer`.

The caller just close the producer, no need to care it's asynchronously or not.

File: cdc/sinkv2/ddlsink/mq/mq_ddl_sink.go (hunk 4 above)
Comment (3AceShowHand) [reply cdc/sinkv2/ddlsink/mq/mq_ddl_sink.go:184]: It's unnecessary to report errors on closing to the caller, just print the log is enough.

No problems like goroutine leak.

File: cdc/sinkv2/eventsink/mq/dmlproducer/kafka_dml_producer.go (hunk 5 above)
Comment (3AceShowHand) [reply cdc/sinkv2/eventsink/mq/dmlproducer/kafka_dml_producer.go:75]: It's moved to `saramaAsyncProducer.Close`

**Review by zhaoxinyu — APPROVED** [review]

**Pull request merged**